### ISSUE TITLE
aztk cluster provision fails during bootstrap because the connection to azure.archive.ubuntu.com:80 times out.

### ISSUE THREAD

**Issue by grahamsk** [Azure/aztk#493]
Hi Guys,

Over the past few days we have been experiencing intermittent issues when launching aztk clusters. Looking at the stdout of the Azure Batch VMs it appears that whilst the VM is bootstrapping itself and running apt-get install {package}, it could not connect to azure.archive.ubuntu.com:80 (51.137.52.58) as the connection timed out.

I realise that this is an Azure /VM/Ubuntu mirror issue as you use the default ubuntu 16.04 image (https://github.com/Azure/aztk/blob/44a07654aaf85dd60804fd632b2c4c6eec8d1b52/aztk/spark/client.py#L44) but was wondering whether there are any plans to allow end users to provide a custom VM image? If not, would there be any interest?

In our particular case I would find it useful to do this as I'd like to add additional mirrors to the VMs sources.list file.

As an alternative I suppose that I could add mirrors directly in this bootstrap script above the highlighted line - https://github.com/Azure/aztk/blob/61e7c591cdfc1586f323d49d736f05883f550cc4/aztk/node_scripts/setup_node.sh#L12

**Comment by grahamsk** [comment]
FYI - I'm in the process of notifying Canonical about the error with azure.archive.ubuntu.com as well.

**Comment by paselem** [comment]
Thanks @grahamsk  for bringing this up. At the moment we have not really considered custom images since our main strategy for customization is around containers. That said, it is a feature the underlying service supports so we could potentially try it out if this becomes more critical. The API change is reasonably trivial, the hard part is making it easy to do.

Adding mirrors is also an interesting option, or perhaps even more control around what gets installed on the host, although there are parts of this where we need to make sure we bootstrap correctly before an end user gets to play around.

What mirrors were you considering adding?

**Comment by jafreck** [comment]
@grahamsk I have not been able to reproduce this issue. Can you give some more details about your setup? Are you using a vnet or do you have any firewall rules that might affect connectivity?

As @paselem said, the VM is really not used in aztk except as a host for the containers -- the container is where the customization comes in. So there are no plans to allow user exposed VM image control as that could affect/break the aztk start scripts.

**Comment by jafreck** [comment]
Closing due to inactivity and inability to reproduce.

**Comment by grahamsk** [comment]
Hey Guys,

Sorry for the delay in response @paselem & @jafreck  - The issue was with azure.archive.ubuntu.com. I contacted mirrors@ubuntu.com who confirmed that the issue was with the mirror and was fixed shortly before you tried to reproduce. It feels like this type of event will probably quite a rare occurence and as such it doesn't warrant much consideration as I guess the container is the main point of focus

If I do think of any cases where being able to modify the underlying host I will contact you guys on Teams.
